strawberry/extensions/runner.py

```diff
@@ -21,37 +27,59 @@ def __init__(
 
     def _run_on_all_extensions(self, method_name: str, *args, **kwargs):
         for extension in self.extensions:
-            getattr(extension, method_name)(*args, **kwargs)
+            method = getattr(extension, method_name)
 
-    @contextmanager
-    def request(self):
-        self._run_on_all_extensions("on_request_start")
+            if inspect.iscoroutinefunction(method):
+                raise RuntimeError(SYNC_CONTEXT_WARNING)
 
-        yield
+            method(*args, **kwargs)
 
-        self._run_on_all_extensions("on_request_end")
+    async def _run_on_all_extensions_async(self, method_name: str, *args, **kwargs):
+        for extension in self.extensions:
+            result = getattr(extension, method_name)(*args, **kwargs)
 
-    @contextmanager
-    def validation(self):
-        self._run_on_all_extensions("on_validation_start")
+            if inspect.isawaitable(result):
+                await result
 
-        yield
+    def request(self) -> RequestContextManger:
+        return RequestContextManger(
+            run_on_all_extensions=self._run_on_all_extensions,
+            run_on_all_extensions_async=self._run_on_all_extensions_async,
+        )
 
-        self._run_on_all_extensions("on_validation_end")
+    def validation(self) -> ValidationContextManager:
+        return ValidationContextManager(
+            run_on_all_extensions=self._run_on_all_extensions,
+            run_on_all_extensions_async=self._run_on_all_extensions_async,
+        )
 
-    @contextmanager
-    def parsing(self):
-        self._run_on_all_extensions("on_parsing_start")
+    def parsing(self) -> ParsingContextManager:
+        return ParsingContextManager(
+            run_on_all_extensions=self._run_on_all_extensions,
+            run_on_all_extensions_async=self._run_on_all_extensions_async,
+        )
```

Comment:
This seems like a fair bit of boilerplate that may continue to grow. Is there a way to further abstract this?

Reply:
I reduced the boilerplate, have a look :)